README.md
# Gravity Bridge Portal by blockscape

A Flutter web UI for the cosmos gravity bridge. It uses Metamask to sign transfers from ethereum to cosmos and Keplr to sign transfers from cosmos to ethereum. This Flutter project can't be compiled to native apps, as the Chrome wallet extensions are required.

## Building and running

Currently, there are two possible enviroments (configuration) you can use:

```
flutter run web -t main.dart // this will run prod
flutter run web -t main_test_gravity_private.dart // this will run beta/testnet version
```

## Running the portal locally

Right now we need to use flutter 3.0.

To run this app use the command: flutter run --device-id chrome --web-renderer html

This will open the portal in a debug chrome browser. You have to copy the address from the debug browser and paste it into your standard chrome browser (for the extensions).
The portal requires the metamask and keplr extensions. Both extensions must be setup with at least one address.

--web-renderer html is required for the token icons to be displayed (other renderers will cause CORS issues).

### Hot reload

Hot reload works, but on each reload you have to switch to the debug browser window first, to trigger the reload. After that you can refresh the page in your default chrome browser.

## Keplr Signing

If you run into issues with the implementation of the keplr interaction you can always look at cosmjs, and how they handle proto signing. Cosmjs was basically the blueprint for this dart implementation. (https://github.com/cosmos/cosmjs)

## Protobuf

The Keplr extension expects protobuf input. To generate protobuf (if .proto files should chane in the future) you have to use the protoc(https://developers.google.com/protocol-buffers/docs/downloads) tool with the proto dart backend(https://github.com/google/protobuf.dart). Since the dart backend is written in dart, you need to have dart in your Path as well.
The script expects the packend to be installed via

```
$ pub global activate protoc_plugin
```

There are two scripts that help you compile the dart files:

- getGravityProto.sh: this downloads the gravity proto files that are required.
- compileProto.sh: this compiles the downloaded .proto files to dart and saves them to lib/services/gravity/proto (protoc and dart backend required).

## Splashscreen

We are using own logic to show the splash screen at start. Please see the index.html, styles.css where we reference on "splash".

## Useful Resources

### Cosmos RPC

https://docs.tendermint.com/master/rpc/#/Info/tx
https://docs.tendermint.com/master/tendermint-core/using-tendermint.html#formatting

## How to translate

To translate a label of message

## Run extensions in chrome developer mode

Copy next function in `.bash_profile` or `.zshrc` file in your user root:

```bash
function flutter_web_enable_extensions() {
  FLUTTER_SDK_DIR=$@;

  echo "Flutter SDK dir: $FLUTTER_SDK_DIR";

  sed -i '' "s/'--disable-extensions'/\/\/'--disable-extensions'/g" $FLUTTER_SDK_DIR/packages/flutter_tools/lib/src/web/chrome.dart
  rm $FLUTTER_SDK_DIR/bin/cache/flutter_tools.stamp
}
```

Source it (or reopen terminal), then execute: `flutter_web_enable_extensions /path/to/flutter/sdk`

## Adding Configurations

### Adding tokens that can be bridged

Add the token to the GravityTokens list in the keplr config. This will allow the token to be selected for transfer from gravity too ethereum. To get the ibc denom you need the gravity ibc channel and the denom of the token. Then you can execute a query with the gravity binary:

```
 gravity q ibc-transfer denom-hash "transfer/channel-{channel_No}/{denom}"

// e.g gravity q ibc-transfer denom-hash "transfer/channel-79/uscrt"
```

There must be a deployed and registered erc20 contract to do this.

### Adding a Forward chain

Add a Forward Chain configuration to the keplr configuration. Additionally you have to extend the Chain enum to include the new chain, as well es create a chain config and map it to the enum in chain_config.dart file. There is also a mapping from enum to asset. A proposal to enable this must have passed before.

### IBC Connection

To add a IBC connection extend the ibcChains in the keplr config. Channels can be found here: https://github.com/Gravity-Bridge/Gravity-Docs/blob/main/docs/list_of_IBC_channels.md

### Token Registry

If a token is expected to end up on Gravity, please add its info to the currency-registry in the keplr config, so it can be displayed correctly.


assets/CHANGELOG.md
### Version 1.1.33 - 27 January 2023

- Feature: Enabled NYM token for Grav-ETH transfers

### Version 1.1.32 - 26 January 2023

- Feature: Adopted Keplr's new data structure for alternative fees with different tokens
- Feature: Updated Terra-Classic's fees
- Feature: Updated Ki-chain's fees
- Feature: Updated config for Iris
- Feature: Disabled Bostrom and Sifchain because of expired IBC-clients
- Bugfix: Fixed explorer link generation for Terra-Classic transactions and accounts
- Bugfix: Improved balance requests for Regen

### Version 1.1.31 - 20 January 2023

- Bugfix: Disabled connecting to a Canto account through Keplr because of incorrect integration between the two. This lead to transfers from Canto being signed incorrectly and failing.
- Bugfix: Improved balance requests for Evmos and Iris.
- Feature: Small UI improvements.

### Version 1.1.30 - 17 January 2023

- Bugfix: "Recent batches" widget, "Transaction queue" widget, and "Batches" page throw UI error when the Metamask network is not "Ethereum Mainnet"
- Bugfix: Modified error message shown when the bridge fee dollar price can not be requested

### Version 1.1.29 - 13 January 2023

- Feature: CHEQD tokens are enabled for transfers to Ethereum
- Feature: Added hint about the special nature of Evmos tokens
- Feature: Disabled ledger connecting for chains that have incompatible bip44 numbers
- Feature: Added alerts about updated fees
- Bugfix: Transactions Gravity-to-Canto are unsuccesful
- Feature: Added alert, which notifies the user that the Canto-bugfix will lead to Keplr generating a new (and empty) Canto account
- Bugfix: Evmos balance requests throw error
- Bugfix: Correct number formatting of chain fees
- Bugfix: Small UI simplifications

### Version 1.1.28 - 30 December 2022

- Bugfix: Fix Luna Classic configurations to enable IBC transfers

### Version 1.1.27 - 15 December 2022

- Bugfix: Amounts in "Recent batches" widget and "Batches" page have sometimes a rounding error
- Bugfix: Widgets are grey when the gravity API returns an error
- Feature: Format numbers according to default browser language
- Feature: Added "Transaction rejected" dialog for Ledger
- Bugfix: Small UI fixes and simplifications in "Batches" page

### Version 1.1.26 - 09 December 2022

- Feature: Ledger support for Cosmos accounts
- Feature: Gravity chain transaction queue widget created
- Bugfix: Batches with tokens not from Uniswap show wrong amount formating and no token name
- Bugfix: Bridge fee selector appears blank after switching transfer route from GRAV-ETH to GRAV-CHI to GRAV-ETH
- Bugfix: Paddings fixed and scrollbars added

### Version 1.1.25 - 02 December 2022

- Feature: Updated gas prices for Chihuahua
- Feature: Added notice about the recent Chihuahua gas changes
- Bugfix: Show an error dialog if cancellation of a transaction fails
- Bugfix: Gravity supply info and volume info widgets toggle size fix
- Bugfix: Some small UI bugs fixed & small improvements

### Version 1.1.24 - 25 November 2022

- Feature: Animated progress icons
- Feature: Number formatting depending on the locale
- Bugfix: Token dialog overflows with very long token names
- Bugfix: Token dialog renders icons incorrectly for unknown tokens

### Version 1.1.23 - 18 November 2022

- Feature: Add FUND token
- Feature: Gravity supply info and volume info widgets created
- Feature: Redirect links added to recent batches and recent transactions widgets rows
- Bugfix: Rounding error when sending the maximally available tokens
- Bugfix: Mobile/desktop form factor edited and UI adjusted
- Bugfix: Popup menu re-ordered

### Version 1.1.22 - 14 November 2022

- Bugfix: Transaction fees for Gravity-ETH transactions disabled

### Version 1.1.21 - 11 November 2022

- Feature: Add transaction fee to Gravity-ETH transactions
- Feature: Show block progress for unfinished ETH-Cosmos transactions
- Feature: Show date and time outputs in local date-time format
- Feature: Transactions and batches pages include links to blockchain accounts
- Bugfix: Improved remaining time calculation in batches view
- Bugfix: When batches still loading, show shimmer effect for recent batches widget and batches page
- Bugfix: Transaction history page UI improvements

### Version 1.1.20 - 04 November 2022

- Feature: Bridge fee selector widget redesign for mobile and desktop views
- Bugfix: Amount text-field warning messages do not show in direction ETH-GRA
- Bugfix: Refresh remaining time of a batch
- Bugfix: Remaining batch time fix

### Version 1.1.19 - 28 October 2022

- Feature: Show batch started time as smalltext
- Feature: Adapted dialogs for slow and standard batch transfers
- Feature: Transfer and batch statistics panels created for mobile and desktop views
- Feature: Transaction history: show remaining time until batch completion
- Bugfix: Connection buttons appear connected despite login in Metamask/Keplr being interrupted
- Bugfix: Re-added token denominations in transaction history
- Bugfix: Fixed card switching inconsistent bug in mobile view of transfer page

### Version 1.1.18.1 - 24 October 2022

- Bugfix: Transaction history throws an error in a special case with old data

### Version 1.1.18 - 21 October 2022

- Feature: Batches monitoring for mobile and desktop view
- Feature: Transaction history redesign for mobile and desktop view
- Feature: Added button to import a Cosmos token into Metamask
- Bugfix: Cosmos-chains whose clients are inactive have been disabled
- Feature: Adapted price and time constraints for the standard batched transfer

### Version 1.1.17 - 10 October 2022

- Bugfix: Connect buttons not being clicked issue fixed
- Bugfix: Popup menu adjustment for desktop view

### Version 1.1.16 - 07 October 2022

- Redesign: Popup menu and entries
- Bugfix: Improved error dialog about wrong Metamask network

### Version 1.1.15.1 - 07 October 2022

- Bugfix: Going back to the home-page reloads it unnecessarily

### Version 1.1.15 - 06 October 2022


lib/main.dart
// /*
//     Gravity Bridge is a UI for interacting with the gravity chain.
//     Copyright (C) 2023  blockscape Finance AG, Switzerland
//
//     This program is free software: you can redistribute it and/or modify
//     it under the terms of the GNU General Public License as published by
//     the Free Software Foundation, either version 3 of the License, or
//     (at your option) any later version.
//
//     This program is distributed in the hope that it will be useful,
//     but WITHOUT ANY WARRANTY; without even the implied warranty of
//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//     GNU General Public License for more details.
//
//     You should have received a copy of the GNU General Public License
//     along with this program.  If not, see <https://www.gnu.org/licenses/>.
// */

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:gravity_bridge_flutter/bin/config.dart';

import 'package:gravity_bridge_flutter/services/chain_configs/chain_configs.dart';
import 'package:gravity_bridge_flutter/services/currency_registry/cr_mainnet.dart';
import 'package:gravity_bridge_flutter/services/currency_registry/currency_registry.dart';
import 'package:url_strategy/url_strategy.dart';

import 'package:gravity_bridge_flutter/app.dart';

class KeplrProd implements KeplrConfig {
  @override
  final String rpc =
      "https://node.gravity-bridge-3-1.fullnodes.location3.blockscape.network/rpc";
  @override
  final String lcd =
      "https://node.gravity-bridge-3-1.fullnodes.location3.blockscape.network/lightclient";
  @override
  final String webGRPC = "https://gravity-webgrpc-1.blockscape.network:443";
  @override
  final String chainID = "gravity-bridge-3";
  @override
  final String bech32BasePrefix = "gravity";
  @override
  final String coinDenomStake = "GRAV";
  @override
  final String coinDenomStakeMin = "ugraviton";
  @override
  final String coinDenomFee = "GRAV";
  @override
  final String coinDenomFeeMin = "ugraviton";
  @override
  final List<GravityToken> gravityTokens = const [
    // GravityToken(           // NOT READY YET: NO ERC20 CONTRACT DEPLOYED
    //   coinGeckoName: "certik",
    //   decimals: 6,
    //   name: "Shentu",
    //   symbol: "CTK",
    //   denom:
    //       "ibc/1250DF5455D8208B8E23C2DB1BD7B6D90CC898F4E3D8708F229AF5547E6674A6",
    // ),
    GravityToken(
      coinGeckoName: "nym",
      decimals: 6,
      name: "Nym",
      symbol: "NYM",
      denom:
          "ibc/0C273962C274B2C05B22D9474BFE5B84D6A6FCAD198CB9B0ACD35EA521A36606",
    ),
    GravityToken(
      coinGeckoName: "cosmos",
      decimals: 6,
      name: "Atom",
      symbol: "ATOM",
      denom:
          "ibc/2E5D0AC026AC1AFA65A23023BA4F24BB8DDF94F118EDC0BAD6F625BFC557CDED",
    ),
    GravityToken(
      coinGeckoName: "comdex",
      decimals: 6,
      name: "Comdex",
      symbol: "CMDX",
      denom:
          "ibc/29A7122D024B5B8FA8A2EFBB4FA47272C25C8926AA005A96807127208082DAB3",
    ),
    GravityToken(
      coinGeckoName: "stargaze",
      decimals: 6,
      name: "Stargaze Stars",
      symbol: "STARS",
      denom:
          "ibc/4F393C3FCA4190C0A6756CE7F6D897D5D1BE57D6CCB80D0BC87393566A7B6602",
    ),
    GravityToken(
      coinGeckoName: "chihuahua-token",
      decimals: 6,
      name: "Chihuahua",
      symbol: "HUAHUA",
      denom:
          "ibc/048BE20AE2E6BFD4142C547E04F17E5F94363003A12B7B6C084E08101BFCF7D1",
    ),
    GravityToken(
      coinGeckoName: "secret",
      decimals: 6,
      name: "Secret",
      symbol: "SCRT",
      denom:
          "ibc/7907EA1A11FD4FC2A815FCAA54948C42F08E3F3C874EE48861386286FEB80160",
    ),
    GravityToken(
      coinGeckoName: "assetmantle",
      decimals: 6,
      name: "AssetMantle",
      symbol: "MNTL",
      denom:
          "ibc/00F2B62EB069321A454B708876476AFCD9C23C8C9C4A5A206DDF1CD96B645057",
    ),
    GravityToken(
        coinGeckoName: "unification",
        decimals: 9,
        name: "FUND",
        symbol: "FUND",
        denom:
            "ibc/D157AD8A50DAB0FC4EB95BBE1D9407A590FA2CDEE04C90A76C005089BF76E519"),
    GravityToken(
        coinGeckoName: "terrausd",
        decimals: 6,
        name: "UST Classic",
        symbol: "USTC",
        denom:
            "ibc/50896BE248180B0341B4A679CF49249ECF70032AF1307BFAF0233D35F0D25665"),
    GravityToken(
        coinGeckoName: "terra-luna",
        decimals: 6,
        name: "Luna Classic",
        symbol: "LUNC",
        denom:
            "ibc/7F04F5D641285808E1C6A01D266C3D9BE1C23473BF3D01AC31E621CFF72DBF24"),
    // GravityToken(
    //   coinGeckoName: "agoric",
    //   decimals: 6,
    //   name: "Agoric",
    //   symbol: "BLD",
    //   denom: "", // denom is missing for Agoric
    // ),
    GravityToken(
      coinGeckoName: "cheqd-network",
      decimals: 9,
      name: "cheqd",
      symbol: "CHEQ",
      denom:
          "ibc/5012B1C96F286E8A6604A87037CE51241C6F1CA195B71D1E261FCACB69FB6BC2",
    ),
    // ),
    // GravityToken( // NOT READY YET: NO ERC20 CONTRACT DEPLOYED
    //   coinGeckoName: "graviton",
    //   decimals: 9,
    //   name: "Graviton",
    //   symbol: "GRAV",
    //   denom: "graviton",
    //   // gravityIbcChannelId: "",
    //   // nativeIbcChannelId: ""),
    // ),
  ];

  @override
  final List<ForwardChain> forwardChains = const [
    ForwardChain(
      addressPrefix: "gravity",
      chainName: Chain.gravity,
      label: "Gravity Bridge",
      isDefault: true,
    ),
    ForwardChain(
      addressPrefix: "n1",
      chainName: Chain.nyx,
      label: "Nyx",
    ),
    ForwardChain(
      addressPrefix: "cosmos1",
      chainName: Chain.cosmosHub,
      label: "Cosmos Hub",
    ),
    ForwardChain(
      addressPrefix: "osmo1",
      chainName: Chain.osmosis,
      label: "Osmosis",
    ),
    ForwardChain(
      addressPrefix: "iaa1",
      chainName: Chain.iris,
      label: "Iris",
    ),
    ForwardChain(
      addressPrefix: "cre1",
      chainName: Chain.crescent,
      label: "Crescent",
    ),
    ForwardChain(
      addressPrefix: "chihuahua1",
      chainName: Chain.chihuahua,
      label: "Chihuahua",
    ),
    ForwardChain(
      addressPrefix: "stars1",
      chainName: Chain.stargaze,
      label: "Stargaze",
    ),
    ForwardChain(
      addressPrefix: "secret",
      chainName: Chain.secret,
      label: "Secret",
    ),
    ForwardChain(
      addressPrefix: "canto",
      chainName: Chain.canto,
      label: "Canto",
    ),
    ForwardChain(
      addressPrefix: "mantle",
      chainName: Chain.assetmantle,
      label: "AssetMantle",
    ),
    ForwardChain(
      addressPrefix: "evmos1",
      chainName: Chain.evmos,
      label: "Evmos",
    ),
    ForwardChain(
      addressPrefix: "regen",
      chainName: Chain.regen,
      label: "Regen",
    ),
    ForwardChain(
      addressPrefix: "agoric",
      chainName: Chain.agoric,
      label: "Agoric",
    ),
    ForwardChain(
      addressPrefix: "und",
      chainName: Chain.unification,
      label: "Unification",
    ),
    ForwardChain(
      addressPrefix: "terra",
      chainName: Chain.terraclassic,
      label: "Terra Classic",
    ),
  ];

  @override
  final List<IbcChain> ibcChains = const [
    IbcChain(
      addressPrefix: "chihuahua1",
      chainName: Chain.chihuahua,
      channelNative: "channel-15",
      channelGravity: "channel-34",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "cosmos1",
      chainName: Chain.cosmosHub,
      channelNative: "channel-281",
      channelGravity: "channel-17",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "cre1",
      chainName: Chain.crescent,
      channelNative: "channel-2",
      channelGravity: "channel-62",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "iaa1",
      chainName: Chain.iris,
      channelNative: "channel-29",
      channelGravity: "channel-47",
      channelID: "transfer",
    ),
    // IbcChain(
    //   addressPrefix: "n1",
    //   chainName: Chain.nyx,
    //   channelNative: "channel-0",
    //   channelGravity: "channel-18",
    //   channelID: "transfer",
    // ),
    IbcChain(
      addressPrefix: "osmo1",
      chainName: Chain.osmosis,
      channelNative: "channel-144",
      channelGravity: "channel-10",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "stars1",
      chainName: Chain.stargaze,
      channelNative: "channel-6",
      channelGravity: "channel-7",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "bcna",
      chainName: Chain.bitcanna,
      channelNative: "channel-8",
      channelGravity: "channel-0",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "juno",
      chainName: Chain.juno,
      channelNative: "channel-31",
      channelGravity: "channel-8",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "certik",
      chainName: Chain.shentu,
      channelNative: "channel-6",
      channelGravity: "channel-12",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "persistence",
      chainName: Chain.persistence,
      channelNative: "channel-38",
      channelGravity: "channel-24",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "bostrom",
      chainName: Chain.bostrom,
      channelNative: "channel-6",
      channelGravity: "channel-33",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "ki",
      chainName: Chain.kichain,
      channelNative: "channel-11",
      channelGravity: "channel-37",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "cheqd",
      chainName: Chain.cheqd,
      channelNative: "channel-2",
      channelGravity: "channel-43",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "terra",
      chainName: Chain.terraclassic,
      channelNative: "channel-64",
      channelGravity: "channel-101",
      channelID: "transfer",
    ),
    IbcChain(
        addressPrefix: "umee",
        chainName: Chain.umee,
        channelNative: "channel-9",
        channelGravity: "channel-28",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "ixo",
        chainName: Chain.impacthub,
        channelNative: "channel-16",
        channelGravity: "channel-26",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "fetch",
        chainName: Chain.fetchai,
        channelNative: "channel-7",
        channelGravity: "channel-30",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "rizon",
        chainName: Chain.rizon,
        channelNative: "channel-4",
        channelGravity: "channel-53",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "lum",
        chainName: Chain.lum,
        channelNative: "channel-6",
        channelGravity: "channel-55",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "sif",
        chainName: Chain.sifchain,
        channelNative: "channel-60",
        channelGravity: "channel-66",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "darc",
        chainName: Chain.konstellation,
        channelNative: "channel-2",
        channelGravity: "channel-57",
        channelID: "transfer"),
    IbcChain(
        addressPrefix: "evmos",
        chainName: Chain.evmos,
        channelNative: "channel-8",
        channelGravity: "channel-65",
        channelID: "transfer"),
    IbcChain(
      addressPrefix: "secret",
      chainName: Chain.secret,
      channelNative: "channel-17",
      channelGravity: "channel-79",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "mantle",
      chainName: Chain.assetmantle,
      channelNative: "channel-8",
      channelGravity: "channel-82",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "canto",
      chainName: Chain.canto,
      channelNative: "channel-0",
      channelGravity: "channel-88",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "regen",
      chainName: Chain.regen,
      channelNative: "channel-45",
      channelGravity: "channel-89",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "agoric",
      chainName: Chain.agoric,
      channelNative: "channel-4",
      channelGravity: "channel-91",
      channelID: "transfer",
    ),
    IbcChain(
      addressPrefix: "und",
      chainName: Chain.unification,
      channelNative: "channel-2",
      channelGravity: "channel-98",
      channelID: "transfer",
    ),
  ];

  @override
  final CurrencyRegistry currencyRegistry = mainnetRegistry;

  const KeplrProd();
}

class MetamaskProd implements MetamaskConfig {
  @override
  final String gravityContractAddr =
      '0xa4108aA1Ec4967F8b52220a4f7e94A8201F2D906';
  @override
  final String networkID = "0x1"; // mainnet
  @override
  final String networkName = "Ethereum Mainnet";

  const MetamaskProd();
}

void main() async {
  Config.init(
    keplr: const KeplrProd(),
    metamask: const MetamaskProd(),
    feeServerAddress: "https://price-server.blockscape.network",
  );

  // Here we set the URL strategy for our web app.
  // It is safe to call this function when running on mobile or desktop as well.
  setPathUrlStrategy();
  runApp(const ProviderScope(child: MyApp()));
}
